File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;

class Solution {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        Map<String, Integer> friendMap = new HashMap<>();
        int[] giftDegree = new int[friends.length];
        int[][] giftGraph = new int[friends.length][friends.length];
        
        for(int i = 0; i < friends.length; i++) {
            friendMap.put(friends[i], i);
        }
        
        for(String gift : gifts) {
            String[] name = gift.split(" ");
            giftDegree[friendMap.get(name[0])]++;
            giftDegree[friendMap.get(name[1])]--;
            giftGraph[friendMap.get(name[0])][friendMap.get(name[1])]++;
        }
        
        for(int i = 0; i < friends.length; i++) {
            int num = 0;
            for(int j = 0; j < friends.length; j++) {
                if(i == j) continue;
                
                if(giftGraph[i][j] > giftGraph[j][i] || (giftGraph[i][j] == giftGraph[j][i] && giftDegree[i] > giftDegree[j])) {
                    num++;
                }
            }
            
            answer = Math.max(answer, num);
        }
        
        return answer;
    }
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.py
def solution(friends, gifts):
    friend_arr = {v: i for i, v in enumerate(friends)}
    n = len(friends)
    degree = [0] * n
    answer = [0] * n
    graph = [[0] * n for i in range(n)]
    
    for gift in gifts:
        give, take = gift.split()
        graph[friend_arr[give]][friend_arr[take]] += 1
        
    for i in range(n):
        degree[i] = sum(graph[i]) - sum([g[i] for g in graph])
        
    for i in range(n):
        for j in range(n):
            if graph[i][j] > graph[j][i]:
                answer[i] += 1
            elif graph[i][j] == graph[j][i] and degree[i] > degree[j]:
                answer[i] += 1
    
    return max(answer)